third_party/e2e/nvme_driver.c

```diff
@@ -0,0 +1,178 @@
+#include "nvme_driver.h"
+#include "nvme_internal.h"
+
+#include <linux/bio.h>
+#include <linux/blk-mq.h>
+#include <linux/blk_types.h>
+#include <linux/blkdev.h>
+#include <linux/completion.h>
+#include <linux/fs.h>
+#include <linux/genhd.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/nvme.h>
+#include <linux/nvme_ioctl.h>
+
+#define MY_BDEV_MODE (FMODE_READ | FMODE_WRITE)
+
+#define BITS_PER_SLICE 6
+#define BITS_PER_WU 7
+#define BITS_PER_DIE 6
+
+struct block_device* bdev;
+struct gendisk* bd_disk;
+struct nvme_ns* ns;
+
+struct nvme_request {
+  struct nvme_command* cmd;
+  union nvme_result result;
+  u8 retries;
+  u8 flags;
+  u16 status;
+  void* ctrl;
+};
+
+static inline struct nvme_request* nvme_req(struct request* req) {
+  return blk_mq_rq_to_pdu(req);
+}
+
+static void submit_req_done(struct request* rq) {
+  struct nvme_request* nvme_rq = nvme_req(rq);
+
+  if (rq) {
+    blk_mq_free_request(rq);
+  }
+}
+
+struct request* nvme_alloc_request(struct request_queue* q,
+                                   struct nvme_command* cmd) {
```

Comment:
here, names like `cmd` and `op` are in general ok because they are widely used/understood.